### ISSUE TITLE
UnitCrossProduct in math_utils is badly implemented

### ISSUE THREAD

**Issue by RiccardoRossi** [KratosMultiphysics/Kratos#664]
the problem here is that

https://github.com/KratosMultiphysics/Kratos/blame/7c6860b8bcafb200ad94cb83b9cf7445616367c6/kratos/utilities/math_utils.h#L455

changes the order of the input arguments.

thing is that  instead of having
AxB = UnitCrossProduct(A,B)

what is currently implemented is
BxA = UnitCrossProduct(A,B)

which obviously changes the sign. I believe this behaviour is sort of evil...so **i suggest changing it even if it breaks the current behaviour**. Can everyone take a look at own code to check for usages? (by grep i saw that the number of usages is limited)

also it would be good that in the math utils, the method should be templated, something like

        template< Tout, TA, TB >CrossProduct(Tout& out , const TA& A, const TB& B)
        {
               out = ...
        }
which would highly reduce code duplication. Not to say that making the output unit is just one line. Do we really need a different function from CrossProduct just for that?

**Comment by loumalouomega** [comment]
Indeed, I use it a lot because is cleaner

**Comment by maceligueta** [comment]
I reported this some months back... 
http://kratos-wiki.cimne.upc.edu/forum/index.php?PHPSESSID=9ocjm8qk7ro60bkn2nkqvahlp3&topic=117.0

**Comment by pooyan-dadvand** [comment]
And I was convinced that it was solved by @loumalouomega by its changes in math_utils.h

**Comment by loumalouomega** [comment]
Maybe there was a discussion and we decided to keep as it was

**Comment by maceligueta** [comment]
Someone said that we could simply remove the methods:

```
static inline array_1d<double, 3> UnitCrossProduct( const array_1d<double, 3>& vec, 
 const array_1d<double, 3>& Tuple) 
{

        array_1d<double, 3> cross;
        cross[0] =  Tuple[1]*vec[2] - Tuple[2]*vec[1];
        cross[1] =  Tuple[2]*vec[0] - Tuple[0]*vec[2];
        cross[2] =  Tuple[0]*vec[1] - Tuple[1]*vec[0];

        const double length = std::sqrt(inner_prod(cross, cross));
        cross = (1.00/length) * cross;
        return cross;
    }

    static inline array_1d<double, 3> CrossProduct(
        const array_1d<double, 3>& vec, 
        const array_1d<double, 3>& Tuple
        )
    {
        array_1d<double, 3> cross;

        cross[0] =  Tuple[1]*vec[2] - Tuple[2]*vec[1];
        cross[1] =  Tuple[2]*vec[0] - Tuple[0]*vec[2];
        cross[2] =  Tuple[0]*vec[1] - Tuple[1]*vec[0];
 
	return cross;
    }
```
    
I think it was @RiccardoRossi .
Because they are slow (temporal arrays are created when returning the output?).
There's also something good in removing them: the error will be explicit. If we just change the order in the product, the kratos users will see a change in the results, and a less evindent reason why it is doing so.

**Comment by RiccardoRossi** [comment]
Hi @maceligueta i confirm what you say about the lack of performance of the methods. Also aesthetically...why vec & tuple???

+1 for removing them completely

**Comment by pooyan-dadvand** [comment]
👍 for removing and 👍 for not changing the order.

**Comment by loumalouomega** [comment]
Do we solved this?

**Comment by maceligueta** [comment]
No, I just checked the master and the problem is still there, everything that was commented.

**Comment by RiccardoRossi** [comment]
so my proposal is the following:

1 - define a Function

       template< T1, T2, T3 >
       CrossProduct(T1& a, const T2& b, const T3& c )   this return a = b x c

2 - remove all of the functions that actually return a value (which are less efficient)
3 - change the signature of the function UnitCrossProduct to also be UnitCrossProduct(a,b,c)

       template< T1, T2, T3 >
       UnitCrossProduct(T1& a, const T2& b, const T3& c )   this return a = b x c / ||b x c ||

those changes deliberately break existing code. This is so as we need to manually ensure what is being done, since right now some functions give a = b x c others (the ones to be removed) give a = c x b

**Comment by loumalouomega** [comment]
The template arguments are array_1d or vectors?

**Comment by RiccardoRossi** [comment]
wrong question.
               the idea is exactly that the T1 T2 T3 can be a combination of arrays and vectors

**Comment by loumalouomega** [comment]
That's what I said

**Comment by pooyan-dadvand** [comment]
:+1: for @RiccardoRossi solution. Only one detail, I'm not sure if is better to start with one template argument or directly go to the generic case.

**Comment by maceligueta** [comment]
I would leave the wrong functions for some time, throwing an explicative error, so coders are aware that the arguments might have a wrong order and that the function is deprecated.

**Comment by pooyan-dadvand** [comment]
I agree that the @maceligueta solution is less intrusive. But I think a compiler error is more future proof than having a code that in a given circumstance stops working due to this error.

**Comment by RiccardoRossi** [comment]
+1,

i will do a PR with this. However i think it should be merged AFTER the release

**Comment by pooyan-dadvand** [comment]
I also think that we need some time to see the lateral effect of this change...

**Comment by RiccardoRossi** [comment]
ok, i will take out the tag release then

**Comment by RiccardoRossi** [comment]
while digging for a solution i found this, discussing on the possibility of defining as "Deprecated" a given function. I think this would be the best option. Shame that we are not in c++14 :-(

https://stackoverflow.com/questions/295120/c-mark-as-deprecated

**Comment by RiccardoRossi** [comment]
would we agree about defining the deprecated macro in the define.h?

**Comment by loumalouomega** [comment]
When will we move to c++14?, in 2 years?

**Comment by RiccardoRossi** [comment]
a good question, to which i have no answer.

**Comment by pooyan-dadvand** [comment]
Not a bad idea to implement this macro! :+1:

**Comment by roigcarlo** [comment]
I'll close this since has already been implemented in #946.
